### PR TITLE
Add hack launcher to the desktop app grid

### PR DESCRIPTION
**Pull request by danigm**
This patch adds a new permanent launcher to the app grid to
enable/disable the HACK mode.

Right now the hack mode enables/disables the codeview and flip to hack
button and the clubhouse component.

https://phabricator.endlessm.com/T27140

### CODE REVIEW

File: js/ui/sessionMode.js [hunk 1]
```diff
@@ -111,11 +111,11 @@ const _modes = {
                     ['networkAgent', 'polkitAgent', 'telepathyClient',
                      'keyring', 'autorunManager', 'automountManager',
                      'appStore', 'discoveryFeed', 'trayArea',
-                     'updaterManager', 'clubhouse'] :
+                     'updaterManager'] :
```
Comment (cfoch) [inline js/ui/sessionMode.js]: why did you remove this? This makes the Clubhouse button to be hidden always.

Comment (danigm) [reply js/ui/sessionMode.js]: I removed that because the clubhouse will be launched by the new desktop icon and not by the gnome-shell. But now I think that removing this breaks the clubhouse notifications and more, so I think I need to work on that a bit more to fix that overrides.

File: js/ui/appDisplay.js [hunk 2]
```diff
@@ -2743,3 +2771,64 @@ class AppCenterIcon extends AppIcon {
...
+        this._iconContainer.add_child(this.icon.actor);
+
+        this.actor.set_child(this._iconContainer);
+
+        this.actor.connect('destroy', this._onDestroy.bind(this));
+    }
+
+    _createIcon(iconSize) {
+        let iconUri = 'resource:///org/gnome/shell/theme/hack-button.svg';
+        if (this._iconState === HackAppIconState.DEACTIVATED)
```
Comment (cfoch) [inline js/ui/appDisplay.js]: `this.iconState` instead of `this._iconState`

Comment (danigm) [reply js/ui/appDisplay.js]: It's the same, because the `iconState` is a getter for the `_iconState`, but yes, we should use the getter instead the attribute itself, I'll change that.

File: js/ui/sessionMode.js (hunk 1 above)
Comment (danigm) [reply js/ui/sessionMode.js]: I've left the clubhouse component and removed all the button code because we won't need that anymore, clubhouse button is not needed anymore.

File: js/ui/appDisplay.js [hunk 3]
```diff
@@ -626,6 +633,10 @@ var AllView = class AllView extends BaseAppView {
     getLayoutIds() {
         let layoutIds = super.getLayoutIds();
 
+        // Only show the hack icon if the clubhouse app is in the system
+        if (_getClubhouse())
+            layoutIds.splice(0, 0, HACK_APP_ID);
```
Comment (erikos) [inline js/ui/appDisplay.js]: This could cause side effects. If an element is not found an empty array is returned. No items on the desktop? :package: 
I think we need to guard this.

Comment (erikos) [reply js/ui/appDisplay.js]: At least when installing/uninstalling the Clubhouse I have seen traces for this in the log and empty desktop grids.

Comment (danigm) [reply js/ui/appDisplay.js]: I think that the problem was not there. I've been reviewing this and I've detected that the `hackAppIcon` variable was not set to null on redraw, causing a crash. I've fixed that and tested installing and uninstalling the clubhouse and now it seems to work correctly.

Comment (erikos) [reply js/ui/appDisplay.js]: Great!
